### PR TITLE
DB usage for Totem/PPS timing local reconstruction

### PR DESCRIPTION
**Pull request by forthommel**
Following the merge of #25777, this PR introduces the new CondFormat for timing calibration for the local reconstruction algorithms for vertical (Totem UFSD) and horizontal (PPS diamond) timing detectors. Currently, only the latter is being matched and used in the PPS offline reconstruction.
Additionally, a few `get` prefixes were stripped from the `TotemTimingConversions` accessors, in preparation for the future CMS coding conventions matchings to the PPS code (see #25690 and CTPPS#40).

The test in `RecoCTPPS/TotemRPLocal/test/totemTiming_reco_cfg.py` has been adapted to allow the user to switch between the condDB retrieval of these calibrations, the use of a external JSON, or the loading of a SQLite file.

cc: @nminafra

### CODE REVIEW

**Comment by perrotta** [comment]
please test
(testing the version without the DB by now: this PR has to be updated, see description)

**Comment by forthommel** [comment]
Indeed, I introduced this workaround for our test, but not for the standard reconstruction sequence... Commit on the way!

File: RecoCTPPS/TotemRPLocal/python/totemTimingLocalReconstruction_cff.py [hunk 1]
```diff
@@ -10,3 +10,4 @@
     totemTimingRecHits *
     totemTimingLocalTracks
 )
+
```
Comment (slava77) [inline RecoCTPPS/TotemRPLocal/python/totemTimingLocalReconstruction_cff.py]: this looks unnecessary

**Comment by slava77** [comment]
> Currently, the new payload has not been uploaded to the conditions DB

when should we expect this to happen?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
It is now in `oracle://cms_orcoff_prep/CMS_CONDITIONS`! But I now need to figure out how to test everything together...

**Comment by boudoul** [comment]
@tocheng , @ggovi , maybe you could  help @forthommel in this process. Thank you much.

File: RecoCTPPS/Configuration/python/recoCTPPS_sequences_cff.py [hunk 2]
```diff
@@ -9,6 +9,11 @@
 from Geometry.VeryForwardGeometryBuilder.ctppsIncludeAlignmentsFromXML_cfi import *
 ctppsIncludeAlignmentsFromXML.RealFiles = cms.vstring("Alignment/CTPPS/data/RPixGeometryCorrections.xml")
 
+#--- temporary workaround while calibration conditions are integrated within the DB
+from CondFormats.CTPPSReadoutObjects.ppsTimingCalibrationESSource_cfi import ppsTimingCalibrationESSource
+ppsTimingCalibrationESSource.calibrationFile = cms.FileInPath('RecoCTPPS/TotemRPLocal/data/timing_offsets_ufsd_2018.dec18.cal.json')
```
Comment (slava77) [inline RecoCTPPS/Configuration/python/recoCTPPS_sequences_cff.py]: similar to what's in L10, this approach of loading a default config and modifying it is dangerous.
The relevant configs for the ES sources should be defined in a single place where such sources are defined.
Any other client of these ES sources is likely to get inconsistent setup.

**Comment by tocheng** [comment]
@forthommel @slava77 @boudoul 
The PPS timing correction tag is not in the autoCond GTs for CMSSW relval workflows.
Before updating the GTs, I'd like to see some validation plots to show the timing correction works as expected using the workaround in the last commitment.
Thanks.
@ggovi What do you think?

File: RecoCTPPS/TotemRPLocal/python/totemTimingLocalReconstruction_cff.py (hunk 1 above)
Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/python/totemTimingLocalReconstruction_cff.py]: Indeed, leftovers from some vim hooks... will remove in a later commit/rebase!

**Comment by ggovi** [comment]
@forthommel @tocheng 
since the payload is already in the prep db, you might run the workflow to consume it by adding a toGet statement pointing to the new tag. Does this help?

**Comment by slava77** [comment]
@forthommel @ggovi 
do you know if an updated global tag for this payload is coming soon?

**Comment by ggovi** [comment]
@tocheng might be able to answer?

**Comment by forthommel** [comment]
@slava77, all,
Sorry for the long radio silence! Just to let you know that, following an offline thread we had with @tocheng et al. we have now uploaded our [calibration payload](https://github.com/cms-sw/cmssw/pull/25777) on the [production DB](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/TotemTimingCalibration_v1). We are now waiting for a new GT to be issued before one may proceed with the testing of this PR.

**Comment by forthommel** [comment]
Please do not test before the new GT is released: @ggovi, @tocheng ?

**Comment by forthommel** [comment]
> Please do not test before the new GT is released: @ggovi, @tocheng ?

Depends on #25929 .

**Comment by forthommel** [comment]
Now that #25929 is merged I think we are ready for testing !

**Comment by perrotta** [comment]
please test

**Comment by forthommel** [comment]
Seems that for [140.56 step2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25858/33160/runTheMatrix-results/140.56_RunHI2018+RunHI2018+RECOHID18+HARVESTDHI18/step2_RAW2DIGI_L1Reco_RECO_ALCA_SKIM_EI_DQM.py) the GT is hardcoded (`103X_dataRun2_Prompt_v2`). But what concerns me more is that the TOTEM timing detectors reconstruction even seems to be triggered on MC validations... Let me investigate!

**Comment by forthommel** [comment]
@perrotta, @slava77, as a follow-up from my [previous comment](https://github.com/cms-sw/cmssw/pull/25858#issuecomment-464327948), is there any reason why some hardcoded GT values are to be found in the testing FW?
I am also still puzzled why the local reco would suddenly be run on MC workflows. It seems that our pixels' local reco (another PPS offline reco code with ES retrieval from DB: see e.g. [here](https://github.com/cms-sw/cmssw/blob/master/RecoCTPPS/PixelLocal/plugins/CTPPSPixelClusterProducer.cc#L41)) is not facing this (even though the check for a `DetSetVector<CTPPSPixelDigi>` might skip this error...)
What do you think?

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc [hunk 3]
```diff
@@ -59,6 +63,13 @@ TotemTimingRecHitProducer::produce( edm::Event& iEvent, const edm::EventSetup& i
 {
   std::unique_ptr<edm::DetSetVector<TotemTimingRecHit> > pOut( new edm::DetSetVector<TotemTimingRecHit> );
 
+  // check for timing calibration parameters update
+  if ( calibWatcher_.check( iSetup ) ) {
+    edm::ESHandle<PPSTimingCalibration> hTimingCalib;
+    iSetup.get<PPSTimingCalibrationRcd>().get( hTimingCalib );
+    algo_.setCalibration( *hTimingCalib );
```
Comment (slava77) [inline RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc]: one possible optimization that could also currently remove the need to have the payloads for MC is to first check the digis in the event record. If it is empty, there is no need to load the ES products or even call the algo.

**Comment by slava77** [comment]
> @perrotta, @slava77, as a follow-up from my [previous comment](https://github.com/cms-sw/cmssw/pull/25858#issuecomment-464327948), is there any reason why some hardcoded GT values are to be found in the testing FW?

This is a problem in  https://github.com/cms-sw/cmssw/blob/807c62dce721302628702aef88362ab404dc5c89/Configuration/PyReleaseValidation/python/relval_steps.py#L1789 it should use ```auto:run2_data_promptlike_hi``` instead of 103X_dataRun2_Prompt_v2

> I am also still puzzled why the local reco would suddenly be run on MC workflows. It seems that our pixels' local reco (another PPS offline reco code with ES retrieval from DB: see e.g. [here](https://github.com/cms-sw/cmssw/blob/master/RecoCTPPS/PixelLocal/plugins/CTPPSPixelClusterProducer.cc#L41)) is not facing this (even though the check for a `DetSetVector<CTPPSPixelDigi>` might skip this error...)
> What do you think?

I think that it would be more practical to add DB payloads to the MC GTs.
We have CTPPS reco enabled starting from the 2016 era.

File: RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc (hunk 3 above)
Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc]: @slava77 Indeed, this is one possible (and working as previously tested) workaround! But I found it personally more confusing (and possibly more prone to false negative on data for incorrect streams) to propagate this check for the validity of the various input collections in all producers downstream. What do you think?

**Comment by forthommel** [comment]
> I think that it would be more practical to add DB payloads to the MC GTs.
> We have CTPPS reco enabled starting from the 2016 era.
I would have guessed the other direction (disable the local reco on MC while we indeed converge on any producers for the digi-level MC), but of course it seems we are more an organisational/conventional matter.
@jan-kaspar, @fabferro, @tocheng, @ggovi ?

**Comment by jan-kaspar** [comment]
> > I think that it would be more practical to add DB payloads to the MC GTs.
> > We have CTPPS reco enabled starting from the 2016 era.
> 
> I would have guessed the other direction (disable the local reco on MC while we indeed converge on any producers for the digi-level MC), but of course it seems we are more an organisational/conventional matter.
> @jan-kaspar, @fabferro, @tocheng, @ggovi ?

We already have disabled PPS reco for some MC workflows:
https://github.com/cms-sw/cmssw/blob/89571866f9da53d669a6a2e86b7214d0be5f874a/Configuration/StandardSequences/python/Reconstruction_cff.py#L93
So maybe we just need to adjust the list of excluded workflows ??

**Comment by slava77** [comment]
On 2/26/19 2:45 AM, jan-kaspar wrote:
> I would have guessed the other direction (disable the local reco on MC
> while we indeed converge on any producers for the digi-level MC), but of
> course it seems we are more an organisational/conventional matter.

PPS simulation was mentioned/promised since 2016 when we started
integrating PPS reco.
So, it was integrated under assumption that simulation for it will
appear "soon".
I seem to recall some progress in the simulation implementation recently.
If I misunderstood, and there will never ever be a simulation of PPS,
then we can consider adjusting the workflows.

**Comment by fabferro** [comment]
@slava77 Hi Slava, PPS simulation is going to appear. The PR you have seem took long to be approved, but it's assumed to be just the first. I've been promised by the simu developers that a new one is going to be issued soon, since the code is there but it needs some adptation to the latest release.

**Comment by forthommel** [comment]
Many thanks, @slava77, @fabferro ! f006eb14352eedfc4b9d725c36f413e9abfd4e8c is now ensuring digis are present before retrieving the calibration payload from condDB.

**Comment by perrotta** [comment]
@cmsbuild please test

File: RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc (hunk 3 above)
Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/plugins/TotemTimingRecHitProducer.cc]: Implemented in f006eb14352eedfc4b9d725c36f413e9abfd4e8c

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25858	f006eb1
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences as expected

**Comment by fabiocos** [comment]
+1

the provided local test configuration runs smoothly

**Pull request merged**